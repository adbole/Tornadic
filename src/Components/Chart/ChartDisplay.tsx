import { AreaChart, BarChart, LineChart } from "recharts";
import { CategoricalChartProps } from "recharts/types/chart/generateCategoricalChart";

import { ChartViews } from ".";


type ChartDisplayProps = {
    children: React.ReactNode;
    property: ChartViews;
} & CategoricalChartProps;

//Provides the proper chart to display based on the HourlyProperty passed
export default function ChartDisplay(props: ChartDisplayProps) {
    const { children, property, ...excess } = props;

    let ChosenChart = BarChart;

    switch (property) {
        case "precipitation":
            ChosenChart = BarChart;
            break;
        case "temperature_2m":
        case "uv_index":
        case "us_aqi":
            ChosenChart = AreaChart;
            break;
        default:
            ChosenChart = LineChart;
    }

    return <ChosenChart {...excess}>{children}</ChosenChart>;
}
