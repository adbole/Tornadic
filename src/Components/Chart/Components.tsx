import { AreaChart, BarChart, LineChart, TooltipProps } from "recharts";
import { CategoricalChartProps } from "recharts/types/chart/generateCategoricalChart";
import { NameType } from "recharts/types/component/DefaultTooltipContent";

import { UV_MAX_VALUES } from "ts/Constants";

import { ChartViews, DataPoint } from ".";

function getUnit(data: DataPoint, defaultUnit: React.ReactNode) {
    const { property, primaryKey } = data;

    if(property === ChartViews.UV_Index) {
        if (primaryKey <= UV_MAX_VALUES.LOW) return " Low"; 
        else if (primaryKey <= UV_MAX_VALUES.MODERATE) return " Moderate";
        else if (primaryKey <= UV_MAX_VALUES.HIGH) return " High";
        else if (primaryKey <= UV_MAX_VALUES.VERY_HIGH) return " Very High";
        else return " Extreme"; 
    }

    return defaultUnit;
}
  
//Customized tooltip to display primary and secondary data with labels if applicable
export const CustomTooltip = ({ active, payload }: TooltipProps<number, NameType>) => {
    if(!active || !payload || !payload.length) return null;

    const fixDecimal = (property: ChartViews, value: number): string | number => 
        (property === ChartViews.Precipitation || property === ChartViews.Pressure) ? value.toFixed(2) : value.toFixed(0);

    const data: DataPoint = payload[0].payload;
    let unit = getUnit(data, payload[0].unit);
    const label = (data.property === ChartViews.Temperature && "Feels: ") || (data.property === ChartViews.Windspeed && "Gust: ");


    return (
        <div className="chart-tooltip">
            <h1>{fixDecimal(data.property, data.primaryKey)}{unit}</h1>
            {data.secondaryKey && <p>{label}{fixDecimal(data.property, data.secondaryKey)}{unit}</p>}
        </div>
    );
};

type ChartDisplayProps = {
    children: React.ReactNode,
    property: ChartViews
} & CategoricalChartProps

//Provides the proper chart to display based on the HourlyProperty passed
export const ChartDisplay = (props: ChartDisplayProps) => {
    const { children, property, ...excess } = props;

    let ChosenChart = BarChart;

    switch(property) {
        case ChartViews.Precipitation:
            ChosenChart = BarChart;
            break;
        case ChartViews.Temperature:
        case ChartViews.UV_Index:
            ChosenChart = AreaChart;
            break;
        default:
            ChosenChart = LineChart;
    }

    return (
        <ChosenChart {...excess}>
            {children}
        </ChosenChart>
    );
};