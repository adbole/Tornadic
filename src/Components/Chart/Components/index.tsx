export { default as Area } from "./Area"
export { default as Axes } from "./Axes"
export { default as Bar } from "./Bar"
export { default as Line } from "./Line"
export { default as NowReference } from "./NowReference"
export { default as Tooltip, useTooltip } from "./Tooltip"