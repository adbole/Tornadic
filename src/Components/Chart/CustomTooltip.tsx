import { TooltipProps } from "recharts";
import { NameType } from "recharts/types/component/DefaultTooltipContent";

import { get_aq, get_uv } from "ts/Helpers";

import { ChartViews, DataPoint } from ".";


function getUnit(data: DataPoint, defaultUnit: React.ReactNode) {
    const { property, primaryKey } = data;

    if(property === "uv_index") {
        return " " + get_uv(primaryKey);
    }
    else if(property === "us_aqi") {
        return " " + get_aq(primaryKey);
    }

    return defaultUnit;
}

//Customized tooltip to display primary and secondary data with labels if applicable
const CustomTooltip = ({ active, payload }: TooltipProps<number, NameType>) => {
    if(!active || !payload || !payload.length) return null;

    const fixDecimal = (property: ChartViews, value: number): string | number => 
        (property === "precipitation" || property === "surface_pressure") ? value.toFixed(2) : value.toFixed(0);

    const data: DataPoint = payload[0].payload;
    const unit = getUnit(data, payload[0].unit);
    const label = 
        (data.property === "temperature_2m" && "Feels: ") || 
        (data.property === "windspeed_10m" && "Gust: ");

    return (
        <div className="chart-tooltip">
            <h1>{fixDecimal(data.property, data.primaryKey)}{unit}</h1>
            {data.secondaryKey && <p>{label}{fixDecimal(data.property, data.secondaryKey)}{unit}</p>}
        </div>
    );
};

export default CustomTooltip;