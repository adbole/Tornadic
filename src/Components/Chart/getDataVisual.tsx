import { Area, Bar, Line } from "recharts";

import { get_aq, get_uv, nameof, toHSL } from "ts/Helpers";

import { ChartViews, DataPoint } from ".";


export default function getDataVisual(unit: string, view: ChartViews, dataPoints: DataPoint[]) {
    switch(view) {
        case "precipitation":
            return <Bar dataKey={nameof<DataPoint>("primaryKey")} fill={"#0078ef"} unit={unit}/>;
        case "temperature_2m": {
            const dataValues = dataPoints.flatMap(point => [point.primaryKey, (point.secondaryKey as number) ?? 0]);
            const min = Math.min(...dataValues);
            const max = Math.max(...dataValues);

            return (
                <>
                    <defs>
                        <linearGradient id="tempGradient" x1="0" y1="0" x2="0" y2="1">
                            <stop offset="0%" stopColor={toHSL(Math.min(120, max))}/>
                            <stop offset="100%" stopColor={toHSL(Math.max(0, min))}/>
                        </linearGradient>
                    </defs>
                    <Area type="monotone" dataKey={nameof<DataPoint>("primaryKey")} stroke="#ffffff00" fillOpacity={0.75} fill="url(#tempGradient)" unit={unit}/>
                    <Area type="monotone" dataKey={nameof<DataPoint>("secondaryKey")} stroke="#fff" fillOpacity={0}/>
                </>
            );
        }
        case "uv_index": {
            const maxUV = Math.max(...dataPoints.map(point => point.primaryKey));
            const maxLevel = get_uv(maxUV);

            return (
                <>
                    <defs>
                        <linearGradient id="tempGradient" x1="0" y1="0" x2="0" y2="1">
                            {maxLevel === "Extreme" && <stop offset="0%" stopColor="#FF00D6"/>}
                            {maxLevel === "Very High" && <stop offset="25%" stopColor="#FF2204"/>}
                            {maxLevel === "High" && <stop offset="50%" stopColor="#FF9431"/>}
                            {maxLevel === "Moderate" && <stop offset="75%" stopColor="#FFF501"/>}
                            <stop offset="100%" stopColor="#00FF66"/>
                        </linearGradient>
                    </defs>
                    <Area type="monotone" dataKey={nameof<DataPoint>("primaryKey")} stroke="#ffffff00" fillOpacity={0.75} fill="url(#tempGradient)"/>
                </>
            );
        }
        case "us_aqi": {
            const maxAQ = Math.max(...dataPoints.map(point => point.primaryKey));
            const maxLevel = get_aq(maxAQ);

            return (
                <>
                    <defs>
                        <linearGradient id="tempGradient" x1="0" y1="0" x2="0" y2="1">
                            {maxLevel === "Hazardous" && <stop offset="0%" stopColor="#6D0000"/>}
                            {maxLevel === "Very Unhealthy" && <stop offset="20%" stopColor="#8400FF"/>}
                            {maxLevel === "Unhealthy" && <stop offset="40%" stopColor="#FF2204"/>}
                            {maxLevel === "Unhealthy*" && <stop offset="60%" stopColor="#FF9431"/>}
                            {maxLevel === "Moderate" && <stop offset="80%" stopColor="#FFF501"/>}
                            <stop offset="100%" stopColor="#00FF66"/>
                        </linearGradient>
                    </defs>
                    <Area type="monotone" dataKey={nameof<DataPoint>("primaryKey")} stroke="#ffffff00" fillOpacity={0.75} fill="url(#tempGradient)"/>
                </>
            );
        }
        default:
            return (
                <>
                    <Line type="monotone" dataKey={nameof<DataPoint>("primaryKey")} stroke="#2668f7" unit={unit}/>
                    {dataPoints[0].secondaryKey != null && <Line type="monotone" dataKey={nameof<DataPoint>("secondaryKey")} stroke="#2eff7d"/>}
                </>
            );
    }
}