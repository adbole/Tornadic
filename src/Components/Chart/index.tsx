import React from "react";
import styled from "@emotion/styled";
import * as d3 from "d3";

import { throwError } from "ts/Helpers";
import { vars } from "ts/StyleMixins";


export type ChartType = "band" | "linear"

export type DataPoint = {
    x: Date;
    y: number[]
};

const ChartContext = React.createContext<{
    chart: d3.Selection<SVGSVGElement, unknown, null, undefined>;
    x: d3.ScaleTime<number, number, never> | d3.ScaleBand<Date>;
    y: d3.ScaleLinear<number, number, never>;
    dataPoints: DataPoint[];
    type: ChartType
} | null>(null);

export const margin = {
    top: 120,
    right: 10,
    bottom: 40,
    left: 40,
};

export const useChart = () =>
    React.useContext(ChartContext) ??
    throwError("useChart must be used within a ChartContextProvider");

const ResponsiveSVG = styled.svg({
    position: "relative",
    border: "1px solid #ffffff19",
    borderRadius: vars.borderRadius,
    overflow: "hidden",
    width: "100%",
    height: "100%",
    display: "block",
});

const defaultYBounds = ([min, max]: [number, number]): [number, number] => {
    const range = max - min;
    const rangePadding = range * 0.1;

    return [min - rangePadding, max + rangePadding];
}

export default function ChartContextProvider({
    dataPoints,
    type,
    children,
    yBounds = defaultYBounds,
}: {
    dataPoints: DataPoint[]
    type: ChartType
    children: React.ReactNode;
    yBounds?: ([min, max]: [number, number]) => [number, number]
}) {
    const [chart, setChart] = React.useState<d3.Selection<
        SVGSVGElement,
        unknown,
        null,
        undefined
    > | null>(null);

    const [width, setWidth] = React.useState<number | null>(null);
    const [height, setHeight] = React.useState<number | null>(null);

    const x = React.useMemo<d3.ScaleTime<number, number> | d3.ScaleBand<Date> | null>(() => {
        if (width === null) return null;

        const range = [margin.left, width - margin.right];

        if (type === "band") {
            return d3
                .scaleBand<Date>()
                .domain(dataPoints.map(point => point.x))
                .range(range)
                .padding(0.1);
        }

        const xDomain = d3.extent(dataPoints, d => d.x) as [Date, Date];

        return d3.scaleTime().domain(xDomain).range(range);
    }, [width, dataPoints, type]);

    const y = React.useMemo<d3.ScaleLinear<number, number> | null>(() => {
        if (height === null) return null;

        const yValues = dataPoints.flatMap(point => point.y);
        const yDomain = yBounds(d3.extent(yValues) as [number, number]);

        return d3
            .scaleLinear()
            .domain(yDomain)
            .range([height - margin.bottom, margin.top]);
    }, [height, dataPoints, yBounds]);

    React.useEffect(() => {
        if (!chart) return;

        const onResize = () => {
            const boundingRect = chart.node()!.getBoundingClientRect();

            const width = boundingRect.width;
            const height = boundingRect.height;

            setWidth(width);
            setHeight(height);

            chart.attr("viewBox", `0 0 ${width ?? 1200} ${height ?? 600}`);
        };

        onResize();

        window.addEventListener("resize", onResize);

        return () => window.removeEventListener("resize", onResize);
    }, [chart]);

    const value = React.useMemo(() => {
        if (!chart || x === null || y === null) return null;

        return {
            chart,
            x,
            y,
            dataPoints,
            type,
        };
    }, [chart, x, y, dataPoints, type]);

    return (
        <ChartContext.Provider value={value}>
            <ResponsiveSVG
                preserveAspectRatio="xMidYMid meet"
                ref={element => {
                    if (!chart && element) setChart(d3.select(element));
                }}
            >
                {value && children}
            </ResponsiveSVG>
        </ChartContext.Provider>
    );
}
