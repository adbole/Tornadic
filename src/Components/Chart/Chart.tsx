import { ResponsiveContainer, Line, CartesianGrid, XAxis, YAxis, Tooltip, Bar, Area, ReferenceLine } from "recharts";
import { CustomTooltip, PressureTick, ChartDisplay } from "./Chart.Components";

import { Forecast, useWeather } from "../Contexes/Weather";
import React from "react";
import { TimeConverter, nameof } from "../../ts/Helpers";
import { Modal, ModalContent, ModalTitle } from "../Contexes/ModalContext";

//Properties denoted here represent values supported for displaying by a Chart
export enum HourlyProperties {
    Temperature = "temperature_2m",
    Humidity = "relativehumidity_2m",
    Precipitation = "precipitation",
    Dewpoint = "dewpoint_2m",
    Visibility = "visibility",
    Windspeed = "windspeed_10m",
    Pressure = "surface_pressure"
}

export type DataPoint = {
    property: HourlyProperties,
    name: string,
    primaryKey: number,
    secondaryKey: number | null
}

const ToHSL = (x: number) => `hsl(${250 * ((120-x)/120)}deg, 100%, 50%)`;

//Gets the data for the given day and property
function GetData(forecastData: Forecast, property: HourlyProperties, day: number) {
    let data: DataPoint[] = [];

    for(let i = 24 * (day); i < 24 * (day + 1); ++i) {
        data.push({
            property: property,
            name: TimeConverter.GetTimeFormatted(forecastData.hourly.time[i], TimeConverter.TimeFormat.Hour),
            primaryKey: forecastData.hourly[property][i],
            secondaryKey: GetSecondaryKey(i)
        });
    }

    return data;

    //Some properties have secondary values
    function GetSecondaryKey(i: number) {
        switch(property) {
            case HourlyProperties.Temperature:
                return forecastData.hourly.apparent_temperature[i];
            case HourlyProperties.Windspeed:
                return forecastData.hourly.windgusts_10m[i];
            default:
                return null;
        }
    }
}

function GetMinMax([min, max]: [number, number], property: HourlyProperties): [number, number] {
    if(property === HourlyProperties.Pressure) {
        min -= 0.1;
        max += 0.1;
    }
    else if(property === HourlyProperties.Precipitation) {
        min = 0;
        max = Math.max(0.5, max);
    }
    else if (property === HourlyProperties.Humidity) {
        min = 0;
        max = 100;
    }
    else {
        min = Math.floor((min * 0.95) / 10) * 10;
        max = Math.ceil((max * 1.10) / 10) * 10;
    }

    return [min, max];
}

function GetDataVisual(forecastData: Forecast, property: HourlyProperties, data: DataPoint[]) {
    switch(property) {
        case HourlyProperties.Precipitation:
            return <Bar dataKey={nameof<DataPoint>("primaryKey")} fill={"#0078ef"} />;
        case HourlyProperties.Temperature:
            const dataValues = data.flatMap(point => [point.primaryKey, (point.secondaryKey as number) ?? 0]);
            const minMax = GetMinMax([Math.min(...dataValues), Math.max(...dataValues)], property);

            return (
                <>
                    <defs>
                        <linearGradient id="tempGradient" x1="0" y1="0" x2="0" y2="1">
                            <stop offset="0%" stopColor={ToHSL(Math.min(120, minMax[1]))} stopOpacity="0.5"/>
                            <stop offset="100%" stopColor={ToHSL(Math.max(0, minMax[0]))} stopOpacity="0.5"/>
                        </linearGradient>
                    </defs>
                    <Area type="monotone" dataKey={nameof<DataPoint>("primaryKey")} stroke="#ffffff00" fillOpacity={1} fill="url(#tempGradient)" unit={forecastData.hourly_units[property]}/>
                    <Area type="monotone" dataKey={nameof<DataPoint>("secondaryKey")} stroke="#fff" fillOpacity={0}/>
                </>
            );
        default:
            return (
                <>
                    <Line type="monotone" dataKey={nameof<DataPoint>("primaryKey")} stroke="#2668f7" unit={forecastData.hourly_units[property]}/>
                    {data[0].secondaryKey != null && <Line type="monotone" dataKey={nameof<DataPoint>("secondaryKey")} stroke="#2eff7d"/>}
                </>
            );
    }
}

const Chart = ({showProperty, showDay = 0}: {showProperty: HourlyProperties, showDay?: number}) => {
    const forecastData = useWeather().forecast;
    const [property, setProperty] = React.useState(showProperty);
    const [day, setDay] = React.useState(showDay);

    const chartData = React.useMemo(() => GetData(forecastData, property, day), [forecastData, property, day]);

    const timeRef = React.useRef<HTMLSpanElement>(null);
    const setTimeText = React.useCallback((s: string) => {
        if(!timeRef.current) return;

        timeRef.current.innerText = s;
    }, []);

    const onMouseMove = React.useCallback((e: any) => e.activeLabel ? setTimeText(", " + e.activeLabel) : setTimeText(""), [setTimeText]);
    const onMouseLeave = React.useCallback(() => setTimeText(""), [setTimeText]);

    return (
       <Modal id="chart">
            <ModalTitle>
                <select className="clear" title="Current Chart" onChange={(e) => setProperty(e.currentTarget.value as HourlyProperties)} value={property}>
                    {
                        Object.keys(HourlyProperties).map(key => (
                            <option key={key} value={HourlyProperties[key as keyof typeof HourlyProperties]}>{key}</option>
                        ))
                    }
                </select>
            </ModalTitle>
            <ModalContent>
                <div className="controls">
                    {
                        forecastData.hourly.time.filter((_, i) =>  i % 24 === 0).map((time, i) => (
                            <div key={i} className="toggle-button" onClick={() => setDay(i)}>
                                <input type="radio" name="chart-radio" id={i.toString()} defaultChecked={i === day}/>
                                <label htmlFor={i.toString()}>{TimeConverter.GetTimeFormatted(time, TimeConverter.TimeFormat.Weekday)}</label>
                            </div>
                        ))
                    }
                </div>

                <p><span>{TimeConverter.GetTimeFormatted(forecastData.hourly.time[day * 24], TimeConverter.TimeFormat.Date)}</span><span ref={timeRef}></span></p>

                <hr></hr>

                <ResponsiveContainer width={"100%"} height="100%">
                    <ChartDisplay property={property} data={chartData} margin={{top: 0, left: 0, right: 0, bottom: 0}} onMouseMove={onMouseMove} onMouseLeave={onMouseLeave}>
                        {GetDataVisual(forecastData, property, chartData)}
                        <CartesianGrid stroke="#ffffff19"/>
                        <XAxis dataKey="name" interval={5} textAnchor="start"/>
                        {
                            property === HourlyProperties.Pressure 
                            ? <YAxis domain={([dataMin, dataMax]) => GetMinMax([dataMin, dataMax], property)} tick={<PressureTick unit={forecastData.hourly_units[property]}/>}/>
                            : <YAxis domain={([dataMin, dataMax]) => GetMinMax([dataMin, dataMax], property)} unit={forecastData.hourly_units[property]} />
                        }
                        
                        {/* @ts-ignore */}
                        <Tooltip wrapperStyle={{ outline: "none" }} position={{x: 'auto', y: 10}} content={<CustomTooltip/>}/>
                        {day === 0 && <ReferenceLine x={TimeConverter.GetTimeFormatted(forecastData.hourly.time[forecastData.nowIndex], TimeConverter.TimeFormat.Hour)}/>}
                    </ChartDisplay>
                </ResponsiveContainer>
            </ModalContent>
       </Modal>
    );
};

export default React.memo(Chart);