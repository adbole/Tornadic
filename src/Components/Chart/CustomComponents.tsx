import { TooltipProps, BarChart, AreaChart, LineChart } from "recharts";
import { CategoricalChartProps } from "recharts/types/chart/generateCategoricalChart";
import { NameType } from "recharts/types/component/DefaultTooltipContent";
import { DataPoint, HourlyProperties } from "./Chart";

//Airpressure needs a special tick in order to display properly
export const PressureTick = ({x, y, payload, unit}: {x: number, y:number, payload: {value: number}, unit: string} & any) => (
    <g transform={`translate(${x},${y})`}>
        <text x={0} y={0} dy={5} fill="#666">
        <tspan textAnchor="end" x="0">{payload.value.toFixed(2)}</tspan>
        <tspan textAnchor="end" x="0" dy="20">{unit}</tspan>
        </text>
    </g>
);
  
//Customized tooltip to display primary and secondary data with labels if applicable
export const CustomTooltip = ({active, payload}: TooltipProps<number, NameType>) => {
    if(!active || !payload || !payload.length) return null;

    const FixDecimal = (property: HourlyProperties, value: number): string | number => 
        (property === HourlyProperties.Precipitation || property === HourlyProperties.Pressure) ? value.toFixed(2) : value.toFixed(0);

    const data: DataPoint = payload[0].payload;
    const unit = payload[0].unit;
    const label = (data.property === HourlyProperties.Temperature && "Feels: ") || (data.property === HourlyProperties.Windspeed && "Gust: ");

    return (
        <div className="chart-tooltip">
            <h1>{FixDecimal(data.property, data.primaryKey)}{unit}</h1>
            {data.secondaryKey && <p>{label}{FixDecimal(data.property, data.secondaryKey)}{unit}</p>}
        </div>
    );
};

type ChartDisplayProps = {
    children: React.ReactNode,
    property: HourlyProperties
} & CategoricalChartProps

//Provides the proper chart to display based on the HourlyProperty passed
export const ChartDisplay = (props: ChartDisplayProps) => {
    const {children, property, ...excess} = props;

    let ChosenChart = BarChart;

    switch(property) {
        case HourlyProperties.Precipitation:
            ChosenChart = BarChart;
            break;
        case HourlyProperties.Temperature:
            ChosenChart = AreaChart;
            break;
        default:
            ChosenChart = LineChart;
    }

    return (
        <ChosenChart {...excess}>
            {children}
        </ChosenChart>
    );
};