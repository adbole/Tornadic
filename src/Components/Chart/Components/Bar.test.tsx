import { render } from "@testing-library/react";
import * as d3 from "d3";

import Chart from "../";

import { Bar } from ".";


vi.mock("d3", async importOriginal => {
    const original = (await importOriginal()) as any;

    return {
        ...original,
        selection: vi.fn(original.selection),
    };
});

const dataPoints = d3.range(24).map(i => ({
    x: new Date(i),
    y: [i],
}));

test("Passes svg props to group", () => {
    const { container } = render(
        <Chart dataPoints={dataPoints} type="band">
            <Bar fill="red" fillOpacity={0.5} />
        </Chart>
    );

    const group = container.querySelector("g");
    expect.soft(group).toHaveAttribute("fill", "red");
    expect.soft(group).toHaveAttribute("fill-opacity", "0.5");
});

test("Renders bars for yIndex = 0 by default", () => {
    const { container } = render(
        <Chart dataPoints={dataPoints} type="band">
            <Bar />
        </Chart>
    );

    expect(container).toMatchSnapshot();
});

test("Render fails if scale isn't a band scale", () => {
    vi.spyOn(console, "error").mockImplementation(() => undefined);

    expect(() =>
        render(
            <Chart dataPoints={dataPoints} type="linear">
                <Bar />
            </Chart>
        )
    ).toThrow();

    vi.mocked(console.error).mockRestore();
});
