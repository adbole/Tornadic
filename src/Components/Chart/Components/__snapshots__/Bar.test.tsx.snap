// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Renders bars for yIndex = 0 by default 1`] = `
.emotion-0 {
  position: relative;
  border: 1px solid #ffffff19;
  border-radius: var(--border-radius);
  overflow: hidden;
  width: 100%;
  height: 100%;
  display: block;
}

<div>
  <svg
    class="emotion-0 emotion-1"
    preserveAspectRatio="xMidYMid meet"
    viewBox="0 0 0 0"
  >
    <g>
      <rect
        width="1.8672199170124484"
        x="37.9253112033195"
        y="-26.666666666666664"
      />
      <rect
        width="1.8672199170124484"
        x="35.850622406639005"
        y="-26.666666666666664"
      />
      <rect
        width="1.8672199170124484"
        x="33.77593360995851"
        y="-26.666666666666664"
      />
      <rect
        width="1.8672199170124484"
        x="31.70124481327801"
        y="-26.666666666666664"
      />
      <rect
        width="1.8672199170124484"
        x="29.62655601659751"
        y="-26.666666666666664"
      />
      <rect
        width="1.8672199170124484"
        x="27.551867219917014"
        y="-26.666666666666664"
      />
      <rect
        width="1.8672199170124484"
        x="25.477178423236516"
        y="-26.666666666666664"
      />
      <rect
        width="1.8672199170124484"
        x="23.40248962655602"
        y="-26.666666666666664"
      />
      <rect
        width="1.8672199170124484"
        x="21.32780082987552"
        y="-26.666666666666664"
      />
      <rect
        width="1.8672199170124484"
        x="19.253112033195023"
        y="-26.666666666666664"
      />
      <rect
        width="1.8672199170124484"
        x="17.178423236514526"
        y="-26.666666666666664"
      />
      <rect
        width="1.8672199170124484"
        x="15.103734439834028"
        y="-26.666666666666664"
      />
      <rect
        width="1.8672199170124484"
        x="13.029045643153527"
        y="-26.666666666666664"
      />
      <rect
        width="1.8672199170124484"
        x="10.95435684647303"
        y="-26.666666666666664"
      />
      <rect
        width="1.8672199170124484"
        x="8.879668049792532"
        y="-26.666666666666664"
      />
      <rect
        width="1.8672199170124484"
        x="6.804979253112034"
        y="-26.666666666666664"
      />
      <rect
        width="1.8672199170124484"
        x="4.730290456431536"
        y="-26.666666666666664"
      />
      <rect
        width="1.8672199170124484"
        x="2.6556016597510386"
        y="-26.666666666666664"
      />
      <rect
        width="1.8672199170124484"
        x="0.5809128630705391"
        y="-26.666666666666664"
      />
      <rect
        width="1.8672199170124484"
        x="-1.4937759336099585"
        y="-26.666666666666664"
      />
      <rect
        width="1.8672199170124484"
        x="-3.568464730290456"
        y="-26.666666666666664"
      />
      <rect
        width="1.8672199170124484"
        x="-5.643153526970955"
        y="-26.666666666666664"
      />
      <rect
        width="1.8672199170124484"
        x="-7.717842323651453"
        y="-26.666666666666664"
      />
      <rect
        width="1.8672199170124484"
        x="-9.792531120331951"
        y="-26.666666666666664"
      />
    </g>
  </svg>
</div>
`;
